### PR TITLE
added RDFValue::operator == (const T&)

### LINKED ISSUE #30: RDFPropertyMap string-conversion
Some types are a bit difficult to handle in the (implementation of the) `RDFPropertyMap`. One of these is `std::string` since it is easily confused with `char*`. Hence there are a few explicit definitions of `RDFValue::operator =` and also a `template <> std::string RDFValue::to(void*) const;` which should handle the conversion to `std::string`.

But the latter is buggy!

```c++
// RDFPropertyMap::Ptr prop = ....

// version 1:
std::string foo = (*prop)["key_to_a_string_value"]; // okay

// version 2:
std::string bar;
bar = (*prop)["key_to_a_string_value"]; // compiler error: ambiguous sth ....
```
I guess the difference between both versions is that the first one uses _copy initialization_ while the second one uses the _assignment operator_... Dunno.

### PR DESCRIPTION
**Pull request by niniemann**
This fixes #30 as good as possible, I think:

The first part of the issue cannot be resolved so easily, since I have no influence on the the assignment operator of `std::string`, which accepts different types, so that the compiler does not know which type it should cast our `RDFValue` to. The solution is to use the `template <class T> RDFValue::get()` method here (which is **not** new, but now works due to the removed _const_):
```c++
std::string foo = map["key"]; // no problem (copy ctor)
// foo = map["key"]; // does not work, ambiguous (assignment operator)
foo = map["key"].get<std::string>(); // works (assignment operator) since we explicitly specify the type to cast to
```

The second part of the issue has to do with comparing values to entries in the `RDFPropertyMap`. I simply added the `template <class T> RDFValue::operator == (const T&)` which casts to `T` and compares the result with the given value. With one exception to correctly compare with `const char*`.

### CODE REVIEW

**Comment by niniemann** [comment]
Sorry, I messed up. This depends on #37 ! Wait for that one to be merged.

**Review by ctieben — COMMENTED** [review]
I will wait on #37 before I review this.
Otherwise its difficult to track to changes between 3 branches (master -> rdfentity-interface -> fix-rdfvalue-casts)

**Comment by niniemann** [comment]
Maybe we should add the other operators, too, before merging this?

**Comment by ctieben** [comment]
Which other operators did you mean in detail?

**Comment by niniemann** [comment]
Comparisons.  Less than, less or equal, greater than, greater or equal.

File: include/sempr/entity/RDFValue.hpp
Comment (ctieben) [inline include/sempr/entity/RDFValue.hpp:213]: Why not const?
Its hard to follow the implicit cast behind this. On the first view its like magic to make `*this` to the type `T`. Could you add a short brief to describe the behavior in behind?

**Comment by ctieben** [comment]
And yes some more operators sounds helpfully and fits in the context of this PR.

File: include/sempr/entity/RDFValue.hpp
Comment (niniemann) [reply include/sempr/entity/RDFValue.hpp:213]: > Why not const?

Because this is just a different way to use the `operator T()`, which is called when trying to cast the RDFValue to something else, and the operator cannot be const (due to the makeValid, I think). I recognized the error when working on the issue, where one solution is to use `get<Sometype>();` when the cast would be ambiguous. So I fixed it and removed the `const`.

> Its hard to follow the implicit cast behind this. On the first view its like magic to make *this to the type T. Could you add a short brief to describe the behavior in behind?

Of course this is magic. The RDFValue is not intended to be used directly, but only in combination with the RDFPropertyMap. The user shall never actually create an instance of it. But he does get an RDFValue whenever he accesses the map, as it is the wrapper for any value and type -- a bit like a QVariant, as is described in the README.

The general thought behind this is:
`template <typename T> operator T()` is used to cast the RDFValue to the actual value the user wanted, and `template <typename T> RDFValue& operator = (const T& other)` allows to assign it _any_ value.

If the cast is ambiguous, use `template <typename T> T get();`

Is this explanaition sufficient? And if yes, which part of it is really missing? ^^

Comment (ctieben) [reply include/sempr/entity/RDFValue.hpp:213]: Yes it is. It helps to point at the other operators that are used implicit. With this you can understand whats going on here with the magic. :-)

Comment (niniemann) [reply include/sempr/entity/RDFValue.hpp:213]: I've added a few more lines to the readme. :)

**Comment by niniemann** [comment]
I've added the other comparison operators, and thought about implementing stuff like incrementation etc., too, but I'm not sure if/how to implement them properly in a generic way. So I'll just leave it as is for now.

File: include/sempr/entity/RDFValue.hpp
Comment (ctieben) [inline include/sempr/entity/RDFValue.hpp]: Do you no any options to avoid the special case for const char* ?

Comment (niniemann) [reply include/sempr/entity/RDFValue.hpp]: There is none. `const char*` is something really special, because it _could_ in principle be the case that you just want to really represent a pointer to a single character -- but it is mostly (?) used as the C-style, '\0'-terminated string we all use. When writing `"Hello"` (e.g. in `m["baz"] == "Hello"`) directly in the code you create a `const char[6]`, and this is the type that gets used in the template if we didn't specify this special case.

Needless to say, trying to cast something to `const char[6]` is bound to fail... ^^

Comment (niniemann) [reply include/sempr/entity/RDFValue.hpp]: Small addition: Of course, if you are willing to use `std::string` _all the time_ we won't need this. But this would mean that you'd have to write:
```c++
if (m["name"] == std::string("Max")) do_something();

// instead of

if (m["name"] == "Max") do_something();
```
If you ask me, I'm happy with the overload :D

Comment (ctieben) [reply include/sempr/entity/RDFValue.hpp]: I expected an answer like this ^^ 
Its fine this way :+1:

**Review by ctieben — APPROVED** [review]

**Pull request merged**
